docs: 2026 roadmap (#4932)
zarf-dev/zarf: site/src/content/docs/roadmap.mdx

## Introduction

This roadmap reflects the work the Zarf maintainers are actively planning and prioritizing as the project moves toward v1.0.0. It is organized by theme rather than strict timeline, though the ordering generally reflects sequencing and dependencies across areas of work.

As always, this is a living document. Community feedback, proposals, and contributions shape how and when items progress. If you have thoughts on any of these areas, we welcome participation in [community meetings](https://zoom-lfx.platform.linuxfoundation.org/meeting/97461829237?password=add48ad5-fc07-4951-96d2-531b72d2a5dc) or through the [Zarf Enhancement Proposals](https://github.com/zarf-dev/proposals) process.

## Development Story

### Schema Evolution (v1beta1)

The next version of the Zarf schema is a foundational milestone on the path to v1.0.0. This work encompasses meaningful structural improvements alongside the deprecations and removals needed to reach a stable, supportable baseline.

Tracked under [#3433](https://github.com/zarf-dev/zarf/issues/3433) 

Key changes include:

> - **New import structures**: Component imports will move to `ZarfComponentConfig`, and skeleton packages will be replaced by remote components.

Reviewer: Maybe a better description will be something along the lines: new zarf.yaml file structure with `PacakgeConfig` and `ComponentConfig` as the top-level kinds. More details in https://github.com/zarf-dev/proposals/blob/main/0048-schema-update-process/README.md

Reviewer 2: That's a fine alternative, though proposal [0051](https://github.com/zarf-dev/proposals/pull/52) is where the new Kinds are introduced

Author: so is the suggestion here to use @soltysh wording but replace the link with the proposal 0051?

Reviewer 2: Yes

Author: OK, done and thanks

> - **New import structures**: Implement a new zarf.yaml file structure, with PacakgeConfig and ComponentConfig as the top-level kinds. More details in [proposals/pull/52](https://github.com/zarf-dev/proposals/pull/52/changes).

Reviewer 2: ```suggestion
- **New import structures**: Implement a new zarf.yaml file structure, with ZarfPackageConfig and ZarfComponentConfig as the top-level kinds. More details in [proposals/pull/52](https://github.com/zarf-dev/proposals/pull/52/changes).
```

> - **Go templating replaces package templates**: Package template variable substitution will be replaced with Go templating and decoupled from the `create` step.

Reviewer: Since we're covering values before, why not drop this item from here? Unless you meant something different.

Reviewer 2: Yeah package templates are replacing the package templating system on create (https://docs.zarf.dev/ref/create/#package-templates). This differs from values on deploy

Author: So do we want to keep this line? I felt like it was different enough to call out but am open to hearing it's not clear.

Reviewer 2: Yeah I would keep it. Packages templates are different than values.

Author: OK, keeping as is then - TY

- **Breaking changes and deprecations**: A full list of changes is tracked in [proposals/pull/52](https://github.com/zarf-dev/proposals/pull/52/changes).

These changes are necessary to reduce accumulated technical debt and provide a schema that can reasonably guarantee backwards compatibility once v1.0.0 ships.

### Values

Zarf Values and Go templating ([#3946](https://github.com/zarf-dev/zarf/issues/3946)) represent a major shift in how packages are parameterized and configured.  Note that this work replaces the legacy `###ZARF_VAR_*###` variable syntax with a structured, Go-template-based values system. 

**Operator and deployer experience:**

> - **Granular value overrides via `-set-values`** ([#4224](https://github.com/zarf-dev/zarf/issues/4224)): Adding a `-set-values` flag so operators can override individual value keys at deploy time using dot-notation (e.g. `-set-values=.registry.image=my-image`), providing Helm-style granularity without requiring a full values file.

Reviewer: Nit: we always use double hyphen with long-name flags, so `--set-values`. Single hyphen are reserved for single letter flags.

Author: not a nit, good catch! my mistake and will fix

> - **Granular value overrides via `--set-values`** ([#4224](https://github.com/zarf-dev/zarf/issues/4224)): Adding a `--set-values` flag so operators can override individual value keys at deploy time using dot-notation (e.g. `-set-values=.registry.image=my-image`), providing Helm-style granularity without requiring a full values file.

Reviewer 2: ```suggestion
- **Granular value overrides via `--set-values`** ([#4224](https://github.com/zarf-dev/zarf/issues/4224)): Adding a `--set-values` flag so operators can override individual value keys at deploy time using dot-notation (e.g. `--set-values=.registry.image=my-image`), providing Helm-style granularity without requiring a full values file.
```

> - **Helm chart full-values passthrough** ([#4630](https://github.com/zarf-dev/zarf/issues/4630)): Improving the experience for package creators who want to allow deployers to override arbitrary Helm chart values without having to define explicit `sourcePath`/`targetPath` mappings for every key.

Reviewer 3: I flagged this issue on the parent and in v1.0.0 in so much as I think it needs documentation but we won't be updating or changing any codified behaviors - I think this line item can be removed.

Reviewer 3: To elaborate - we won't actually provide any fall-values passthrough behaviors beyond what is available currently.

Author: So we should drop line 34 then?

Author: pertaining to #4630

Author: fixed!

> - **Remote values file sources** ([#4223](https://github.com/zarf-dev/zarf/issues/4223)): Enabling values files to be loaded from HTTP/HTTPS URLs or OCI registries, in addition to local filesystem paths.

Reviewer 2: I would remove the remote file source from this list, it's not something I'd consider v1, and we haven't committed to working on it

Author: ok, removing this row

- **Early validation of undefined values** ([#4693](https://github.com/zarf-dev/zarf/issues/4693)): Zarf should detect missing required values before a deployment begins, rather than failing mid-deploy after components have already been applied.

**Package creator experience:**

- **Zarf builtins in template context** ([#4226](https://github.com/zarf-dev/zarf/issues/4226)): Adding a dedicated key (e.g. `.Zarf`) to the Go template context for Zarf-managed builtins like registry credentials and cluster state, rather than mixing them into the `.Variables` map.

> - **Expose `ZARF_TEMP` in actions** ([#4219](https://github.com/zarf-dev/zarf/issues/4219)): Making the Zarf temporary directory accessible in action contexts via the new templating system (e.g. `{{ .Zarf.tmpDir }}`), eliminating the need for brittle filesystem-search workarounds in airgap deployments.

Reviewer 2: I'd also remove #4219. It is something I'd like to see, but we haven't committed to it and it's a pretty niche use case. IMO not worth calling out in this higher level doc.

Author: roger that, removing

- **Static (non-overridable) values** ([#4878](https://github.com/zarf-dev/zarf/issues/4878)): Allowing package creators to declare values that are locked at create time and cannot be overridden by deployers. Useful for baked-in image references and security-sensitive configuration.

**Schema and composability:**

- **Merge values schema during component import** ([#4877](https://github.com/zarf-dev/zarf/issues/4877)): When a parent package imports a child component, the child’s `values-schema.json` should be merged into the parent’s schema (with the parent winning on conflicts), mirroring how `values.yaml` files are already merged.
- **Skeleton package support** ([#4869](https://github.com/zarf-dev/zarf/issues/4869)): Ensuring skeleton packages correctly copy and resolve values files and schemas, so that packages importing skeletons can use the values feature end-to-end.

> **Init package and migration:**

Reviewer 2: This doesn't render as a header, I think you need an extra newline

Author: good catch, added one above and below. i probably removed something and lost the newline 😄

- **Replace variables in the init package with values** ([#4328](https://github.com/zarf-dev/zarf/issues/4328)): Dogfooding the values feature by migrating the Zarf init package away from legacy variable templating. Care will be taken to avoid collisions with Go templates intended for downstream systems.

> - **Expose init configuration as Zarf variables** ([#1725](https://github.com/zarf-dev/zarf/issues/1725)): Converting init-time flags like `-nodeport` and `-storage-class` to standard Zarf variables so they can be set via `-set` and through the package deploy path. Backwards compatibility with existing flags will be maintained.

Reviewer 2: ```suggestion
- **Expose init configuration as Zarf Values** ([#1725](https://github.com/zarf-dev/zarf/issues/1725)): Converting init-time flags like `--nodeport` and `--storage-class` to standard Zarf values so they can be set via `--set` and through the package deploy path. Backwards compatibility with existing flags will be maintained.
```
Issue originally has this as variables, but it'll end up being implemented through values.

**Documentation:**

- **Clarify values, variables, and value files** ([#4745](https://github.com/zarf-dev/zarf/issues/4745)): The current documentation around values, variables, value files, and Helm chart configuration is a source of significant confusion for new users. This work will produce clearer, more consistent documentation covering where and how templating applies across manifests, charts, actions, and config files.

### Tooling: `zarf tools wait-for`

Two changes are in progress for `zarf tools wait-for`:

- The legacy `zarf tools wait-for` command was deprecated in favor of `zarf tools wait-for resource` and `zarf tools wait-for network`. It will be removed in v1.0.0 ([#4550](https://github.com/zarf-dev/zarf/issues/4550)).

> - `zarf tools wait-for resource` now waits for readiness by default, rather than resource existence ([#4077](https://github.com/zarf-dev/zarf/issues/4077)). This behavior will also be reflected in the v1beta1 schema.

Reviewer 2: ```suggestion
- `zarf tools wait-for resource` now waits for readiness by default, rather than resource existence ([#4077](https://github.com/zarf-dev/zarf/issues/4077)). This behavior will also be reflected when `.cluster.wait` actions are run in v1beta1 packages.
```

### Package Signing

Package signing enhancements ([#4289](https://github.com/zarf-dev/zarf/issues/4289)) cover the full lifecycle of trust for Zarf packages:

> - **Keyless signing support**: Enabling `sigstore` compatible keyless workflows for environments that can support them.

Reviewer 2: ```suggestion
- **Keyless signing support**: Enabling [sigstore](https://www.sigstore.dev/) compatible keyless workflows for environments that can support them.
```

- **SDK deprecation policy** ([#4482](https://github.com/zarf-dev/zarf/issues/4482)): Establishing a clear policy for how SDK-level interfaces are deprecated and removed.
- **Threat model** ([#4813](https://github.com/zarf-dev/zarf/issues/4813)): Documenting Zarf's security architecture to inform both implementation and security review. This includes trust boundaries, sensitive data flows, threat scenarios, and mitigations.
- **RBAC requirements** ([#4722](https://github.com/zarf-dev/zarf/issues/4722)): Documenting the Kubernetes RBAC permissions Zarf requires and why.

### End User Experience

A set of improvements targeting the experience of operators deploying and managing packages:

- **Agent mutation selection** ([#4419](https://github.com/zarf-dev/zarf/issues/4419)): Giving users more control over which resources the Zarf agent mutates during deployment.
- **Connected mode** ([#4580](https://github.com/zarf-dev/zarf/issues/4580)): A `--connected` flag for `zarf package deploy` that allows the same package to be deployed in connected environments without mirroring images or repositories. This also paves the way for removing the `yolo` construct from the v1beta1 schema.

> - **Git fallback** ([#2772](https://github.com/zarf-dev/zarf/issues/2772)): Improving resilience when Git operations encounter errors during package deployment.

Reviewer 2: ```suggestion
- **Git fallback** ([#2772](https://github.com/zarf-dev/zarf/issues/2772)): Improve integrations with Go Git libraries so the fallback shell out to host `git` can be removed.
```
This is for on create not deploy

- **Package state collisions and conflicts** ([#3629](https://github.com/zarf-dev/zarf/issues/3629)): Better detection and handling of conflicting state when multiple packages interact with overlapping resources.

### Security

> Targeted security improvements that complement the broader signing and schema work:

Reviewer 2: I would say that these are not related to the signing and schema work

Author: OK, made this more generic to be just "targeted security improvements"
